src/Microsoft.Framework.DependencyInjection.Abstractions/ServiceCollectionExtensions.cs

```diff
@@ -307,6 +339,15 @@ public static IServiceCollection AddTransient<TService>([NotNull] this IServiceC
             return services.AddTransient(typeof(TService), implementationFactory);
         }
 
+        public static IServiceCollection AddTransient<TService, TImplementation>(
+            [NotNull] this IServiceCollection services,
+            [NotNull] Func<IServiceProvider, TImplementation> implementationFactory)
+            where TService : class
+            where TImplementation : class
```

Comment:
`AddTransient<TService, TImplementation>` has a `where TImplementation : TService` constraint. Maybe we should do the same here. Same for scoped and singleton.

`TryAdd*<TService, TImplementation>` don't seem to have any constraints, so maybe we should add them there too.